model_benchmarking/preprocessing_scripts/label_quailty_check.py

```diff
@@ -0,0 +1,438 @@
+                            os.unlink(file.path)
+                    print("All files deleted successfully.")
+            except OSError:
+                print("No files exist.")
+            
+            # write sampled label to the JSON file
+            file_name = folder_name+'_sampled_animal_count_key.json'
+            file_path = os.path.join(target_folder, file_name)
+                
+            # copy sampled images into the new folder
+            for i in range(len(sampled_images)):
+                file_name = sampled_filenames[i].replace(folder_name+"/", "").replace("/", "_")
+                sampled_images[i]['file_name'] = file_name
+                target_path = target_folder+"/"+file_name
+                shutil.copy(sampled_data_paths[i], target_path)
+            
+            new_coco = {
+                "licenses": coco.dataset.get('licenses', []),
+                "info": coco.dataset.get('info', {}),
+                "categories": coco.dataset.get('categories', []),
+                "images": sorted(sampled_images, key=lambda x: x["id"]),
+                "annotations": sorted(sampled_labels, key=lambda x: x["id"]),
+            }
+            
+            with open(file_path, 'w') as json_file:
+                json.dump(new_coco, json_file, indent=4)
+            
+            print("Number of images sampled: ", len(new_coco["images"]), ' Number of labels sampled: ', len(new_coco["annotations"]))
+                
+        return None
+        
+def iou_for_bboxes(bboxA, bboxB):
+
+    xA_tl, yA_tl, wA, hA = bboxA
+    xB_tl, yB_tl, wB, hB = bboxB
+    
+    xA_br = xA_tl + wA
+    yA_br = yA_tl - hA
+    xB_br = xB_tl + wB
+    yB_br = yB_tl - hB
+    
+    # intersection area
+    xAB_tl = max(xA_tl, xB_tl)
+    yAB_tl = min(yA_tl, yB_tl)
+    xAB_br = min(xA_br, xB_br)
+    yAB_br = max(yA_br, yB_br)
+    w_AB = xAB_br - xAB_tl
+    h_AB = yAB_tl - yAB_br
+    inter_area = w_AB * h_AB
+    if inter_area <= 0:
+        return -1
+    
+    union_area = wA * hA + wB * hB - inter_area
+    
+    iou = inter_area / float(union_area)
+    if iou > 1:
+        return -1
+    
+    return iou
+    
+# Method adapted from: https://gist.github.com/AruniRC/c629c2df0e68e23aff7dcaeef87c72d4 
+def iou_for_multilabels_on_one_image(bboxes_gt_ls, bboxes_pred_ls, iou_limit = 0.5):
+    n_gt = len(bboxes_gt_ls)
+    n_pred = len(bboxes_pred_ls)
+    min_iou = 0.0
+    
+    iou_matrix = np.zeros((n_gt, n_pred))
+    for i in range(n_gt):
+        for j in range(n_pred):
+            iou_matrix[i, j] = iou_for_bboxes(bboxes_gt_ls[i], bboxes_pred_ls[j])
+            
+    # Hungarian matching to find the best matched pairs of old and new labels yielding largest iou
+    idxs_gt, idxs_pred = scipy.optimize.linear_sum_assignment(-iou_matrix)
+    
+    matched_iou = []
+    
+    for i, j in zip(idxs_gt, idxs_pred):
+        matched_iou.append(iou_matrix[i, j])
+    
+    return np.mean(matched_iou)
```

Comment:
Why are we using the mean? How many values are returned from linear_sum_assignment?

Reply:
The number of values returned by linear_sum_assignment corresponds to the number of matched bounding box pairs between the old and new labels. It returns an array of IOU values for each pair of bounding boxes. So, I used np.mean to calculate the final IOU here.